Update required command

yan-soon opened the pull request:
Updated required command to factor in equivalent and preclusion modules.

Notable changes:

**JsonUtil**
- Added new method getPreclusionMapFromJsonFile() that takes in a file's content and converts it into a Map<String, String>.

**RequiredCommandStorage**
- Added a new attribute preclusionMap to store the map precluded modules.
- Added new method getPreclusionMap() to retrieve the preclusionMap.
- Added new method setPreclusionMap() which takes in a file path and uses the getFileFromResource() method and getPreclusionMapFromJsonFile() method to store the given file into the preclusionMap attribute.

**RequiredCommand**
- Added new method isModuleAPreclusion() to check if modules are preclusion, by referencing the preclusionMap.
- Updated the compareModules() method accordingly to include the new method.

Added the relevant tests for the new methods used.

shaokiat commented on src/main/java/seedu/address/logic/commands/RequiredCommand.java:
```diff
@@ -105,11 +129,13 @@ public boolean doesModuleAlreadyExist(Module module, ObservableList<Module> modu
      * @param successMessage Success message for particular category of Modules.
      */
-    public void compareModules(ObservableList<Module> modules, String failMessage, String successMessage) {
+    public void compareModules(ObservableList<Module> modules, Map<String, String> preclusionMap,
+                               String failMessage, String successMessage) {
```
Just curious why did you put `preclusionMap` as a parameter? Can it be stored as a field and accessed through the class instead since there is only 1 preclusionMap. I'm not too sure I'm just asking

yan-soon replied:
Good point, I'm unsure if I should store it as a class variable because idk if the RequiredCommand object should deal with the preclusionMap storage. But it does make the code a bit cleaner if I store it as well HAHAHA

Silvernitro replied:
Actually we don't need to store it separately since we already store `storage`, which contains the map right HAHA

And just to add on, notice that `execute` and `compareModules` are now aware of `preclusionMap`, even though it is only used by `isModuleAPreclusion`, a much lower-level method. The parameter is being passed through 2 methods despite them never using it. This anti-pattern is also called prop drilling/threading in JS btw HAHA

Maybe we can just do `Map<String, String> preclusionMap = storage.getPreclusionMap()` within the `isModuleAPreclusion` fxn instead of in `execute`?

yan-soon replied:
Good point....I totally forgot Im already storing it in storage HAHAH changed accordingly!

shaokiat reviewed (approved):
LGTM! Well elaborated code 💯 👍

Silvernitro commented on src/main/java/seedu/address/logic/commands/RequiredCommand.java:
```diff
@@ -97,6 +100,27 @@ public boolean doesModuleAlreadyExist(Module module, ObservableList<Module> modu
     }
 
+    /**
+     * Checks if a Module is a valid preclusion.
```
Coding standard nit: there should be a blank line after the first line

yan-soon replied:
okay roger! I think i'll open up another PR to fix all the java docs separately to avoid over crowding this one

Silvernitro commented on src/main/java/seedu/address/commons/util/JsonUtil.java:
```diff
@@ -156,4 +159,10 @@ private Level getLoggingLevel(String loggingLevelString) {
     }
+
+    public static Map<String, String> getPreclusionMapFromJsonFile(String file) throws IOException {
+        ObjectMapper mapper = new ObjectMapper();
```
Would it work if we use the mapper that's already defined in the class?

yan-soon replied:
Oh right, it actually does HAHA changed accordingly

Silvernitro commented on src/main/java/seedu/address/logic/commands/RequiredCommand.java:
```diff
@@ -97,6 +100,27 @@ public boolean doesModuleAlreadyExist(Module module, ObservableList<Module> modu
+                                       Map<String, String> preclusionMap) {
+        for (Module mod : modules) {
+            String currentModName = mod.getModuleCode().toString();
+            String modToCheckAgainst = module.getModuleCode().toString();
```
Maybe we can move this out of the loop since it doesn't change every iteration

yan-soon replied:
got it,moved the String modToCheckAgainst outside of the loop

Silvernitro commented on src/main/java/seedu/address/logic/commands/RequiredCommand.java:
```diff
@@ -97,6 +100,27 @@ public boolean doesModuleAlreadyExist(Module module, ObservableList<Module> modu
+        for (Module mod : modules) {
+            String currentModName = mod.getModuleCode().toString();
+            String modToCheckAgainst = module.getModuleCode().toString();
+            if (preclusionMap.containsKey(currentModName)) {
```
We don't actually need to check `containsKey` HAHA if the map doesn't contain the key then the modules won't be equal (the next line)

yan-soon replied:
HAHAHA thanks, changed accordingly

Silvernitro commented on src/main/resources/data/precludedmodules.json:
```diff
@@ -0,0 +1,7 @@
+{
+  "CS1010X" : "CS1101S",
```
I think there are still a lot of preclusions HAHA like cs1010s, cs1010j, etc.

yan-soon replied:
Updated accordingly!

yan-soon commented:
Updates:

Updated code as per team review.

Notable changes:
- Added more preclusion modules.
- Now, only the isModuleAPreclusion() method will deal with the preclusion map.
- To deal with modules that have multiple preclusions, the map now stores a String of all the precluded modules
and in order to check if the module is a preclusion, a contains() check will be done to see if that module is within the
String of all precluded modules.
- Updated javadocs for RequiredCommand, RequiredCommandStorage, GemCommand, GemCommandStorage and ScienceCommand classes. (Added the line in between param and description)
